File: patterns/groups-in-rules/rules-js/groups-in-rules.js
rules.JSRule({
  name: "A Door Sensor Changed",
  triggers: [triggers.GroupStateChangeTrigger("gDoorsSensors")],
  execute: (event) => {
    const door = items.getItem(event.itemName);
    const timer = items.getItem(door.name + "_Timer");
    const lastUpdate = items.getItem(door.name + "_LastUpdate");

    if (door.state === "OPEN") timer.sendCommand("ON");
    else timer.postUpdate("OFF");

    lastUpdate.postUpdate(new Date().toISOString());

    let msg = door.name + (door.state === "OPEN" ? " was opened" : " was closed");

    let alert = false;
    const timeOfDay = items.getItem("vTimeOfDay").state;
    if (timeOfDay === "NIGHT" || timeOfDay === "BED") {
      alert = true;
      msg += " at night";
    }
    if (items.getItem("vPresent").state === "OFF") {
      alert = true;
      msg += " and no one is home";
    }

    if (alert) items.getItem("aAlerts").sendCommand(msg);
    console.log(msg);
  }
});

rules.JSRule({
  name: "Timer expired for a door",
  triggers: [triggers.GroupStateUpdateTrigger("gDoorsTimers", "OFF")],
  execute: (event) => {
    const doorName = event.itemName.split("_")[0];

    const openDoors = items.getItem("gDoorsSensors").members
      .filter((d) => d.state === "OPEN")
      .map((d) => d.name)
      .join(", ");

    let msg = doorName + " has been open for over an hour";
    if (openDoors) msg += " and also open: " + openDoors;

    items.getItem("aAlerts").sendCommand(msg);
    console.log(msg);

    const timeOfDay = items.getItem("vTimeOfDay").state;
    if (timeOfDay === "NIGHT" || timeOfDay === "BED") {
      items.getItem(event.itemName).sendCommand("ON");
    }
  }
});

File: patterns/groups-in-rules/rules-python/groups_in_rules.py
from datetime import datetime

from openhab import rule, Registry
from openhab.triggers import GroupStateChangeTrigger, GroupStateUpdateTrigger


@rule(triggers=[GroupStateChangeTrigger("gDoorsSensors")])
class ADoorSensorChanged:
    def execute(self, module, input):
        event = input.get("event")
        if not event:
            return

        door = Registry.getItem(event.getItemName())
        timer = Registry.getItem(door.getName() + "_Timer")
        last_update = Registry.getItem(door.getName() + "_LastUpdate")

        door_state = str(door.getState())

        if door_state == "OPEN":
            timer.sendCommand("ON")
        else:
            timer.postUpdate("OFF")

        last_update.postUpdate(datetime.now().astimezone())

        msg = door.getName() + (" was opened" if door_state == "OPEN" else " was closed")

        alert = False
        time_of_day = str(Registry.getItem("vTimeOfDay").getState())
        if time_of_day in ("NIGHT", "BED"):
            alert = True
            msg += " at night"
        if str(Registry.getItem("vPresent").getState()) == "OFF":
            alert = True
            msg += " and no one is home"

        if alert:
            Registry.getItem("aAlerts").sendCommand(msg)
        self.logger.info(msg)


@rule(triggers=[GroupStateUpdateTrigger("gDoorsTimers", "OFF")])
class TimerExpiredForADoor:
    def execute(self, module, input):
        event = input.get("event")
        if not event:
            return

        item_name = event.getItemName()
        door_name = item_name.split("_")[0]

        open_doors = [
            d.getName() for d in Registry.getItem("gDoorsSensors").getAllMembers()
            if str(d.getState()) == "OPEN"
        ]

        msg = door_name + " has been open for over an hour"
        if open_doors:
            msg += " and also open: " + ", ".join(open_doors)

        Registry.getItem("aAlerts").sendCommand(msg)
        self.logger.info(msg)

        time_of_day = str(Registry.getItem("vTimeOfDay").getState())
        if time_of_day in ("NIGHT", "BED"):
            Registry.getItem(item_name).sendCommand("ON")
